lib/web/fetch/index.js

```diff
@@ -120,8 +122,49 @@ class Fetch extends EE {
   }
 }
 
+// publishIfSubscribed will check if there is context if so,
+// it will add error in the case of an error before publishing
+// the event if not it will return right away
+function publishIfSubscribed (name, context, error) {
+  if (!context) return
+  if (error) context.error = error
+  channels[name].publish(context)
+}
+
+// handlePromise will handle both fulfilled/rejected promises
+// from p and will publish events accordingly
+function handlePromise (p, context) {
+  return p.promise.then(
+    (success) => {
+      channels.asyncStart.publish(context)
+      channels.asyncEnd.publish(context)
+      return success
+    },
+    (failure) => {
+      context.error = failure
+      channels.asyncStart.publish(context)
+      channels.asyncEnd.publish(context)
+      return Promise.reject(failure)
+    })
+}
```

Comment:
why not the following? 

```suggestion
function handlePromise (p, context) {
  channels.asyncStart.publish(context)
  return p.promise.then(
    (success) => {
      channels.asyncEnd.publish(context)
      return success
    },
    (failure) => {
      context.error = failure
      channels.asyncEnd.publish(context)
      return Promise.reject(failure)
    })
}
```

Reply:
@Uzlopak we still need to call both `asyncStart` and `asyncEnd`.

Reply:
How is my suggestion contradicting?

Reply:
@Uzlopak That would be incorrect timing. The `asyncStart` is supposed to happen when the promise resolves or rejects, not when it's first constructed. The `start` event is what represents the start of the task.